c++/src/kj/compat/gzip.c++

```diff
@@ -226,86 +226,82 @@ Promise<size_t> GzipAsyncInputStream::readImpl(
 
 // =======================================================================================
 
-GzipAsyncOutputStream::GzipAsyncOutputStream(AsyncOutputStream& inner, int compressionLevel)
+GzipAsyncOutputStream::GzipAsyncOutputStream(AsyncOutputStream& inner, kj::Maybe<int> compressionLevel)
     : inner(inner) {
   memset(&ctx, 0, sizeof(ctx));
   ctx.next_in = nullptr;
   ctx.avail_in = 0;
   ctx.next_out = nullptr;
   ctx.avail_out = 0;
 
-  int initResult =
-      deflateInit2(&ctx, compressionLevel, Z_DEFLATED,
+  int initResult;
+  KJ_IF_MAYBE(level, compressionLevel) {
+    compressing = true;
+    initResult =
+      deflateInit2(&ctx, *level, Z_DEFLATED,
                    15 + 16,  // windowBits = 15 (maximum) + magic value 16 to ask for gzip.
                    8,        // memLevel = 8 (the default)
                    Z_DEFAULT_STRATEGY);
-  KJ_ASSERT(initResult == Z_OK, initResult);
+  } else {
+    compressing = false;
+    initResult = inflateInit2(&ctx, 15 + 16);
+  }
+  if (initResult != Z_OK) {
+    fail(initResult);
+  }
 }
 
 GzipAsyncOutputStream::~GzipAsyncOutputStream() noexcept(false) {
-  deflateEnd(&ctx);
+  compressing ? deflateEnd(&ctx) : inflateEnd(&ctx);
 }
 
 Promise<void> GzipAsyncOutputStream::write(const void* in, size_t size) {
   ctx.next_in = const_cast<byte*>(reinterpret_cast<const byte*>(in));
   ctx.avail_in = size;
-  return pump();
+  return pump(Z_NO_FLUSH);
 }
 
 Promise<void> GzipAsyncOutputStream::write(ArrayPtr<const ArrayPtr<const byte>> pieces) {
-  KJ_REQUIRE(!ended, "already ended");
```

Comment:
It performs this check on its own, yeah, I wrote about that in the extended commit message here: https://github.com/capnproto/capnproto/pull/687/commits/cce09b573e5d47248a2d92b15099aa649dc7524d

Haven't checked how reasonable message is though.

Reply:
Checked, message says "stream error". Not extremely descriptive, but seems goon enough for debugging given that it throws as part of the wrong `.write` after `.end` has been already called.

Reply:
Eh... in real code it'll probably be hard to tell that `.end()` had been called and that that's the problem. But I don't feel strongly about this so ok.

Reply:
I'd argue that in real code many cases require more thorough debugging, and, if calling `write` after `end` would be one of the common mistakes, it would likely already have a custom message as part of zlib. Not that I have too strong opinion on this either, just doesn't seem worth a special case at this point.